File: bindings/go/core/compiler.go
// Package core contains the fundamental components of the STRling compiler.
package core

// STRling Compiler - AST to IR Transformation
//
// This module implements the compiler that transforms Abstract Syntax Tree (AST)
// nodes from the parser into an optimized Intermediate Representation (IR). The
// compilation process includes:
//   - Lowering AST nodes to IR operations
//   - Flattening nested sequences and alternations
//   - Coalescing adjacent literal nodes for efficiency
//   - Ensuring quantifier children are properly grouped
//   - Analyzing and tracking regex features used
//
// The IR is designed to be easily consumed by target emitters (e.g., PCRE2)
// while maintaining semantic accuracy and enabling optimizations.

// Compiler transforms AST nodes into optimized IR.
//
// The Compiler class handles the complete transformation pipeline from parsed
// AST to normalized IR, including feature detection for metadata generation.
//
// AST -> IR lowering with normalization:
//   - Flatten nested Seq/Alt
//   - Coalesce adjacent Lit nodes
//   - Ensure quantifier children are grouped appropriately
type Compiler struct {
	featuresUsed map[string]bool
}

// NewCompiler creates a new Compiler instance.
func NewCompiler() *Compiler {
	return &Compiler{
		featuresUsed: make(map[string]bool),
	}
}

// CompileWithMetadata compiles an AST node and returns IR with metadata.
//
// This is the main entry point for compilation with full metadata tracking.
// It performs lowering, normalization, and feature analysis.
func (c *Compiler) CompileWithMetadata(rootNode Node) map[string]interface{} {
	irRoot := c.lower(rootNode)
	irRoot = c.normalize(irRoot)
	
	// Analyze the final IR tree for special features
	c.analyzeFeatures(irRoot)
	
	features := make([]string, 0, len(c.featuresUsed))
	for f := range c.featuresUsed {
		features = append(features, f)
	}
	
	return map[string]interface{}{
		"ir": irRoot,
		"metadata": map[string]interface{}{
			"features_used": features,
		},
	}
}

// Compile compiles an AST node to IR without metadata.
func (c *Compiler) Compile(root Node) IROp {
	ir := c.lower(root)
	ir = c.normalize(ir)
	return ir
}

// analyzeFeatures recursively walks the IR tree and logs features used.
func (c *Compiler) analyzeFeatures(node IROp) {
	switch n := node.(type) {
	case IRGroup:
		if n.Atomic != nil && *n.Atomic {
			c.featuresUsed["atomic_group"] = true
		}
		if n.Name != nil {
			c.featuresUsed["named_group"] = true
		}
		c.analyzeFeatures(n.Body)
		
	case IRQuant:
		if n.Mode == "Possessive" {
			c.featuresUsed["possessive_quantifier"] = true
		}
		c.analyzeFeatures(n.Child)
		
	case IRLook:
		if n.Dir == "Behind" {
			c.featuresUsed["lookbehind"] = true
		} else if n.Dir == "Ahead" {
			c.featuresUsed["lookahead"] = true
		}
		c.analyzeFeatures(n.Body)
		
	case IRBackref:
		c.featuresUsed["backreference"] = true
		
	case IRCharClass:
		// Check for Unicode property escapes in character class items
		for _, item := range n.Items {
			if esc, ok := item.(IRClassEscape); ok && esc.Type == "UnicodeProperty" {
				c.featuresUsed["unicode_property"] = true
			}
		}
		
	case IRSeq:
		for _, part := range n.Parts {
			c.analyzeFeatures(part)
		}
		
	case IRAlt:
		for _, branch := range n.Branches {
			c.analyzeFeatures(branch)
		}
	}
}

// lower converts AST nodes to IR operations.
func (c *Compiler) lower(node Node) IROp {
	switch n := node.(type) {
	case Alt:
		branches := make([]IROp, len(n.Branches))
		for i, b := range n.Branches {
			branches[i] = c.lower(b)
		}
		return IRAlt{Branches: branches}
		
	case Seq:
		parts := make([]IROp, len(n.Parts))
		for i, p := range n.Parts {
			parts[i] = c.lower(p)
		}
		return IRSeq{Parts: parts}
		
	case Lit:
		return IRLit{Value: n.Value}
		
	case Dot:
		return IRDot{}
		
	case Anchor:
		return IRAnchor{At: n.At}
		
	case CharClass:
		items := make([]IRClassItem, len(n.Items))
		for i, item := range n.Items {
			items[i] = c.lowerClassItem(item)
		}
		return IRCharClass{Negated: n.Negated, Items: items}
		
	case Quant:
		child := c.lower(n.Child)
		return IRQuant{Child: child, Min: n.Min, Max: n.Max, Mode: n.Mode}
		
	case Group:
		body := c.lower(n.Body)
		return IRGroup{Capturing: n.Capturing, Body: body, Name: n.Name, Atomic: n.Atomic}
		
	case Backref:
		return IRBackref{ByIndex: n.ByIndex, ByName: n.ByName}
		
	case Look:
		body := c.lower(n.Body)
		return IRLook{Dir: n.Dir, Neg: n.Neg, Body: body}
		
	default:
		// Unknown node type
		return IRLit{Value: ""}
	}
}

// lowerClassItem converts AST class items to IR class items.
func (c *Compiler) lowerClassItem(item ClassItem) IRClassItem {
	switch i := item.(type) {
	case ClassLiteral:
		return IRClassLiteral{Ch: i.Ch}
	case ClassRange:
		return IRClassRange{FromCh: i.FromCh, ToCh: i.ToCh}
	case ClassEscape:
		return IRClassEscape{Type: i.Type, Property: i.Property}
	default:
		return IRClassLiteral{Ch: ""}
	}
}

// normalize optimizes the IR tree.
func (c *Compiler) normalize(node IROp) IROp {
	switch n := node.(type) {
	case IRSeq:
		// Normalize children first
		parts := make([]IROp, len(n.Parts))
		for i, p := range n.Parts {
			parts[i] = c.normalize(p)
		}
		
		// Flatten nested sequences
		flattened := []IROp{}
		for _, part := range parts {
			if seq, ok := part.(IRSeq); ok {
				flattened = append(flattened, seq.Parts...)
			} else {
				flattened = append(flattened, part)
			}
		}
		
		// Coalesce adjacent literals
		coalesced := []IROp{}
		for i := 0; i < len(flattened); i++ {
			if lit, ok := flattened[i].(IRLit); ok {
				// Collect consecutive literals
				value := lit.Value
				j := i + 1
				for j < len(flattened) {
					if nextLit, ok := flattened[j].(IRLit); ok {
						value += nextLit.Value
						j++
					} else {
						break
					}
				}
				coalesced = append(coalesced, IRLit{Value: value})
				i = j - 1
			} else {
				coalesced = append(coalesced, flattened[i])
			}
		}
		
		if len(coalesced) == 1 {
			return coalesced[0]
		}
		return IRSeq{Parts: coalesced}
		
	case IRAlt:
		// Normalize children first
		branches := make([]IROp, len(n.Branches))
		for i, b := range n.Branches {
			branches[i] = c.normalize(b)
		}
		
		// Flatten nested alternations
		flattened := []IROp{}
		for _, branch := range branches {
			if alt, ok := branch.(IRAlt); ok {
				flattened = append(flattened, alt.Branches...)
			} else {
				flattened = append(flattened, branch)
			}
		}
		
		if len(flattened) == 1 {
			return flattened[0]
		}
		return IRAlt{Branches: flattened}
		
	case IRQuant:
		child := c.normalize(n.Child)
		return IRQuant{Child: child, Min: n.Min, Max: n.Max, Mode: n.Mode}
		
	case IRGroup:
		body := c.normalize(n.Body)
		return IRGroup{Capturing: n.Capturing, Body: body, Name: n.Name, Atomic: n.Atomic}
		
	case IRLook:
		body := c.normalize(n.Body)
		return IRLook{Dir: n.Dir, Neg: n.Neg, Body: body}
		
	default:
		return node
	}
}

File: bindings/rust/src/core/compiler.rs
//! STRling Compiler - AST to IR Transformation
//!
//! This module implements the compiler that transforms Abstract Syntax Tree (AST)
//! nodes from the parser into an optimized Intermediate Representation (IR). The
//! compilation process includes:
//!   - Lowering AST nodes to IR operations
//!   - Flattening nested sequences and alternations
//!   - Coalescing adjacent literal nodes for efficiency
//!   - Ensuring quantifier children are properly grouped
//!   - Analyzing and tracking regex features used
//!
//! The IR is designed to be easily consumed by target emitters (e.g., PCRE2)
//! while maintaining semantic accuracy and enabling optimizations.

use crate::core::ir::*;
use crate::core::nodes::*;
use std::collections::HashSet;

/// Compiler for transforming AST nodes into optimized IR.
///
/// The Compiler class handles the complete transformation pipeline from parsed
/// AST to normalized IR, including feature detection for metadata generation.
pub struct Compiler {
    features_used: HashSet<String>,
}

impl Compiler {
    /// Create a new compiler instance
    pub fn new() -> Self {
        Self {
            features_used: HashSet::new(),
        }
    }

    /// Compile an AST node and return IR with metadata
    ///
    /// This is the main entry point for compilation with full metadata tracking.
    /// It performs lowering, normalization, and feature analysis.
    pub fn compile_with_metadata(&mut self, root_node: &Node) -> CompileResult {
        let ir_root = self.lower(root_node);
        let ir_root = self.normalize(ir_root);
        
        self.analyze_features(&ir_root);
        
        CompileResult {
            ir: ir_root,
            metadata: Metadata {
                features_used: self.features_used.iter().cloned().collect(),
            },
        }
    }

    /// Compile an AST node to IR without metadata
    pub fn compile(&mut self, root: &Node) -> IROp {
        let ir = self.lower(root);
        self.normalize(ir)
    }

    /// Lower AST node to IR
    fn lower(&self, node: &Node) -> IROp {
        match node {
            Node::Literal(lit) => IROp::Lit(IRLit {
                value: lit.value.clone(),
            }),
            Node::Dot(_) => IROp::Dot(IRDot {}),
            Node::Anchor(anchor) => {
                let at = if anchor.at == "NonWordBoundary" {
                    "NotWordBoundary".to_string()
                } else {
                    anchor.at.clone()
                };
                IROp::Anchor(IRAnchor { at })
            },
            Node::Sequence(seq) => {
                let parts: Vec<IROp> = seq.parts.iter().map(|p| self.lower(p)).collect();
                IROp::Seq(IRSeq { parts })
            }
            Node::Alternation(alt) => {
                let branches: Vec<IROp> = alt.branches.iter().map(|b| self.lower(b)).collect();
                IROp::Alt(IRAlt { branches })
            }
            Node::Quantifier(quant) => {
                let max = match &quant.max {
                    MaxBound::Finite(n) => IRMaxBound::Finite(*n),
                    MaxBound::Infinite(s) => IRMaxBound::Infinite(s.clone()),
                    MaxBound::Null(_) => IRMaxBound::Infinite("Inf".to_string()),
                };
                
                let mode = if quant.possessive {
                    "Possessive".to_string()
                } else if quant.lazy {
                    "Lazy".to_string()
                } else {
                    "Greedy".to_string()
                };

                IROp::Quant(IRQuant {
                    child: Box::new(self.lower(&quant.target.child)),
                    min: quant.min,
                    max,
                    mode,
                })
            }
            Node::Group(group) => IROp::Group(IRGroup {
                capturing: group.capturing,
                name: group.name.clone(),
                atomic: group.atomic.unwrap_or(false),
                body: Box::new(self.lower(&group.body)),
            }),
            Node::Lookahead(look) => IROp::Look(IRLook {
                dir: "Ahead".to_string(),
                neg: false,
                body: Box::new(self.lower(&look.body)),
            }),
            Node::NegativeLookahead(look) => IROp::Look(IRLook {
                dir: "Ahead".to_string(),
                neg: true,
                body: Box::new(self.lower(&look.body)),
            }),
            Node::Lookbehind(look) => IROp::Look(IRLook {
                dir: "Behind".to_string(),
                neg: false,
                body: Box::new(self.lower(&look.body)),
            }),
            Node::NegativeLookbehind(look) => IROp::Look(IRLook {
                dir: "Behind".to_string(),
                neg: true,
                body: Box::new(self.lower(&look.body)),
            }),
            Node::Backreference(backref) => IROp::Backref(IRBackref {
                by_index: backref.by_index,
                by_name: backref.by_name.clone(),
            }),
            Node::CharacterClass(cc) => IROp::CharClass(IRCharClass {
                negated: cc.negated,
                items: cc.items.iter().map(|item| self.lower_class_item(item)).collect(),
            }),
        }
    }

    /// Lower a class item from AST to IR
    fn lower_class_item(&self, item: &ClassItem) -> IRClassItem {
        match item {
            ClassItem::Char(lit) => IRClassItem::Char(IRClassLiteral {
                ch: lit.ch.clone(),
            }),
            ClassItem::Range(range) => IRClassItem::Range(IRClassRange {
                from_ch: range.from_ch.clone(),
                to_ch: range.to_ch.clone(),
            }),
            ClassItem::Esc(esc) => IRClassItem::Esc(IRClassEscape {
                escape_type: esc.escape_type.clone(),
                property: esc.property.clone(),
            }),
            ClassItem::UnicodeProperty(up) => {
                // Map UnicodeProperty entries into an IR class escape of type 'p'/'P'
                let etype = if up.negated { "P".to_string() } else { "p".to_string() };
                IRClassItem::Esc(IRClassEscape {
                    escape_type: etype,
                    property: Some(up.value.clone()),
                })
            }
        }
    }

    /// Normalize IR (flatten, coalesce, etc.)
    fn normalize(&self, node: IROp) -> IROp {
        match node {
            IROp::Seq(seq) => {
                // Flatten nested sequences
                let mut new_parts = Vec::new();
                for part in seq.parts {
                    let normalized = self.normalize(part);
                    if let IROp::Seq(inner_seq) = normalized {
                        new_parts.extend(inner_seq.parts);
                    } else {
                        new_parts.push(normalized);
                    }
                }
                
                // Coalesce adjacent literals
                let mut coalesced = Vec::new();
                let mut pending_lit = String::new();
                
                for part in new_parts {
                    if let IROp::Lit(lit) = &part {
                        pending_lit.push_str(&lit.value);
                    } else {
                        if !pending_lit.is_empty() {
                            coalesced.push(IROp::Lit(IRLit {
                                value: pending_lit.clone(),
                            }));
                            pending_lit.clear();
                        }
                        coalesced.push(part);
                    }
                }
                
                if !pending_lit.is_empty() {
                    coalesced.push(IROp::Lit(IRLit {
                        value: pending_lit,
                    }));
                }
                
                if coalesced.len() == 1 {
                    coalesced.into_iter().next().unwrap()
                } else {
                    IROp::Seq(IRSeq { parts: coalesced })
                }
            }
            IROp::Alt(mut alt) => {
                // Normalize branches
                alt.branches = alt.branches.into_iter().map(|b| self.normalize(b)).collect();
                IROp::Alt(alt)
            }
            IROp::Quant(mut quant) => {
                quant.child = Box::new(self.normalize(*quant.child));
                IROp::Quant(quant)
            }
            IROp::Group(mut group) => {
                group.body = Box::new(self.normalize(*group.body));
                // Normalize atomic from Option<bool> if needed
                IROp::Group(group)
            }
            IROp::Look(mut look) => {
                look.body = Box::new(self.normalize(*look.body));
                IROp::Look(look)
            }
            other => other,
        }
    }

    /// Analyze IR tree for features used
    fn analyze_features(&mut self, node: &IROp) {
        match node {
            IROp::Group(group) => {
                if group.atomic {
                    self.features_used.insert("atomic_group".to_string());
                }
                if group.name.is_some() {
                    self.features_used.insert("named_group".to_string());
                }
                self.analyze_features(&group.body);
            }
            IROp::Quant(quant) => {
                if quant.mode == "Possessive" {
                    self.features_used.insert("possessive_quantifier".to_string());
                }
                self.analyze_features(&quant.child);
            }
            IROp::Look(look) => {
                if look.dir == "Behind" {
                    self.features_used.insert("lookbehind".to_string());
                } else if look.dir == "Ahead" {
                    self.features_used.insert("lookahead".to_string());
                }
                self.analyze_features(&look.body);
            }
            IROp::Backref(_) => {
                self.features_used.insert("backreference".to_string());
            }
            IROp::CharClass(cc) => {
                for item in &cc.items {
                    if let IRClassItem::Esc(esc) = item {
                        if esc.escape_type == "p" || esc.escape_type == "P" {
                            self.features_used.insert("unicode_property".to_string());
                        }
                    }
                }
            }
            IROp::Seq(seq) => {
                for part in &seq.parts {
                    self.analyze_features(part);
                }
            }
            IROp::Alt(alt) => {
                for branch in &alt.branches {
                    self.analyze_features(branch);
                }
            }
            _ => {}
        }
    }
}

impl Default for Compiler {
    fn default() -> Self {
        Self::new()
    }
}

/// Result of compilation with metadata
#[derive(Debug, Clone)]
pub struct CompileResult {
    pub ir: IROp,
    pub metadata: Metadata,
}

/// Metadata about the compiled pattern
#[derive(Debug, Clone)]
pub struct Metadata {
    pub features_used: Vec<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_compile_literal() {
        let mut compiler = Compiler::new();
        let node = Node::Literal(Literal {
            value: "test".to_string(),
        });
        let ir = compiler.compile(&node);
        match ir {
            IROp::Lit(lit) => assert_eq!(lit.value, "test"),
            _ => panic!("Expected IRLit"),
        }
    }

    #[test]
    fn test_compile_sequence() {
        let mut compiler = Compiler::new();
        let node = Node::Sequence(Sequence {
            parts: vec![
                Node::Literal(Literal {
                    value: "a".to_string(),
                }),
                Node::Literal(Literal {
                    value: "b".to_string(),
                }),
            ],
        });
        let ir = compiler.compile(&node);
        // Should coalesce into a single literal
        match ir {
            IROp::Lit(lit) => assert_eq!(lit.value, "ab"),
            _ => panic!("Expected coalesced literal"),
        }
    }
}
